### PR TITLE
Introduce NDC transaction manager & utils

### PR DESCRIPTION
**Pull request by wxing1292**
* Introduce nDCTransactionMgr which handles backfill of workflow events & create / update transaction redirection
* Introduce nDCCreateTransactionMgr which handles NDC workflow creation scenarios
* Introduce nDCUpdateTransactionMgr which handles NDC workflow update scenarios
* Introduce nDCWorkflow which contains helper functions for NDC transactions
* Update nDCHistoryReplicator accordingly, now nDCHistoryReplicator only handles logics related to replication

### CODE REVIEW

File: service/history/mutableStateBuilder.go [hunk 1]
```diff
@@ -92,9 +92,11 @@ type (
...
 		// in memory only attributes
 		// indicates whether there are buffered events in persistence
-		hasBufferedEventsInPersistence bool
-		// indicates the next event ID in DB, for condition update
-		condition int64
+		hasBufferedEventsInDB bool
+		// indicates whether the workflow is pointed by current record, guaranteed
+		isWorkflowBeingCurrentGuaranteed bool
+		// indicates the next event ID in DB, for conditional update
+		nextEventInDB int64
```
Comment (yux0) [inline service/history/mutableStateBuilder.go]: what about nextEventIDInDB

File: service/history/nDCUpdateTransactionMgr.go [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+	// this is a performance optimization so most update does not need to
+	// check whether target workflow is current workflow by calling DB API
+	if targetWorkflow.getMutableState().IsCurrentWorkflowGuaranteed() {
+		// update to current record, since target workflow is pointed by current record
+		return r.dispatchWorkflowUpdateAsCurrent(
+			ctx,
+			now,
+			isWorkflowRebuilt,
+			targetWorkflow,
+			newWorkflow,
```
Comment (yux0) [inline service/history/nDCUpdateTransactionMgr.go]: Is it only for continue as new?

File: service/history/nDCTransactionMgr.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+		) error
+		backfillWorkflow(
+			ctx ctx.Context,
+			now time.Time,
+			targetWorkflow nDCWorkflow,
+			targetWorkflowEvents *persistence.WorkflowEvents,
+		) error
+	}
+
+	nDCTransactionMgrImpl struct {
```
Comment (yux0) [inline service/history/nDCTransactionMgr.go]: So this is a nested layer like nDCTransactionMgrImpl contains nDCCreateTransactionMgrImpl and nDCCreateTransactionMgrImpl contains nDCTransactionMgrImpl?

File: service/history/nDCCreateTransactionMgr.go [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+	workflowID := targetWorkflow.getMutableState().GetExecutionInfo().WorkflowID
+	targetRunID := targetWorkflow.getMutableState().GetExecutionInfo().RunID
+
+	// we need to check the current workflow execution
+	currentRunID, err := r.transactionMgr.getCurrentWorkflowRunID(
+		ctx,
+		domainID,
+		workflowID,
+	)
+	if err != nil || currentRunID == targetRunID {
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/nDCCreateTransactionMgr.go]: nop, if currentRunID == targetRunID happens, means this replication message is a duplicated message

File: service/history/nDCWorkflow.go [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+
+func (r *nDCWorkflowImpl) terminateWorkflow(
+	lastWriteVersion int64,
+	incomingLastWriteVersion int64,
+) error {
+
+	// TODO nDC: modify for version history
+	// do not persist the change right now, NDC requires transaction
+	r.mutableState.UpdateReplicationStateVersion(lastWriteVersion, true)
+	_, err := r.mutableState.AddWorkflowExecutionTerminatedEvent(
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/nDCWorkflow.go]: good catch

File: service/history/nDCReplicationTask.go [hunk 6]
```diff
@@ -197,7 +203,64 @@ func (t *nDCReplicationTaskImpl) getRequest() *h.ReplicateEventsRequest {
...
+	taskStartTime time.Time,
+) (nDCReplicationTask, error) {
+
+	if len(t.newHistoryEvents) == 0 {
+		return nil, ErrNoNewRunHistory
+	}
+	newHistoryEvents := t.newHistoryEvents
+
+	if t.getLastEvent().GetEventType() != shared.EventTypeWorkflowExecutionContinuedAsNew {
+		return nil, ErrNoNewRunHistory
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/nDCReplicationTask.go]: sure

File: service/history/nDCTransactionMgr.go [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+//  pointed by the current record in DB
+//
+// 2. targetWorkflow means the workflow to be replicated
+//  pointed by the replication task
+//
+// 3. newWorkflow means the workflow to be replicated, as part of continue as new
+//  pointed by the replication task
+//
+// 4. if target workflow and current workflow are the same
+//  then target workflow is set, current workflow is nil
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/nDCTransactionMgr.go]: nop, current workflow being nil meaning that current workflow is not participated in the transaction

File: service/history/nDCCreateTransactionMgr.go [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+}
+
+func (r *nDCCreateTransactionMgrImpl) createAsZombie(
+	ctx ctx.Context,
+	now time.Time,
+	currentWorkflow nDCWorkflow,
+	targetWorkflow nDCWorkflow,
+) error {
+
+	if err := targetWorkflow.suppressWorkflow(
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/nDCCreateTransactionMgr.go]: this means target workflow is suppressed by current workflow

File: service/history/nDCUpdateTransactionMgr.go (hunk 2 above)
Comment (wxing1292) [reply service/history/nDCUpdateTransactionMgr.go]: nop, continue as new is handled as part of applying replication task, except when apply replication task to workflow's non current branch

File: service/history/nDCTransactionMgr.go (hunk 3 above)
Comment (wxing1292) [reply service/history/nDCTransactionMgr.go]: nDCTransactionMgrImpl contains the necessary common methods, for nDCCreateTransactionMgrImpl and nDUpdateTransactionMgrImpl.

nDCTransactionMgrImpl will do dispatching to nDCCreateTransactionMgrImpl and nDUpdateTransactionMgrImpl accordingly.

splitting nDCCreateTransactionMgrImpl and nDUpdateTransactionMgrImpl helps keeping the logics clean

**Pull request merged**